File: src/test/java/com/mmust/TokenizerTest.java
package com.mmust;

import com.mmust.token.NumToken;
import com.mmust.token.Parenthesis;
import com.mmust.tokenizer.Tokenizer;
import org.assertj.core.api.Assertions;
import org.junit.Test;

import static com.mmust.token.Operators.*;

public class TokenizerTest {

	@Test
	public void testArithmeticOperatorsOnly() {
		Tokenizer tokenizer = new Tokenizer("3+4-5*2-9/8");

		Assertions.assertThat(tokenizer)
				.hasSize(11)
				.containsExactly(
						new NumToken("3"),
						PLUS,
						new NumToken("4"),
						MINUS,
						new NumToken("5"),
						MULTIPLY,
						new NumToken("2"),
						MINUS,
						new NumToken("9"),
						DIV,
						new NumToken("8")
				);
	}

	@Test
	public void testWithParenthesis() {
		Tokenizer tokenizer = new Tokenizer("3*(5-1)");

		Assertions.assertThat(tokenizer)
				.hasSize(7)
				.containsExactly(
						new NumToken("3"),
						MULTIPLY,
						Parenthesis.LEFT,
						new NumToken("5"),
						MINUS,
						new NumToken("1"),
						Parenthesis.RIGHT
				);
	}
}

File: src/test/kotlin/com/mmust/tokenizer/TokenizerTest.kt
package com.mmust.tokenizer

import com.mmust.token.ArithmeticOperators.*
import com.mmust.token.LeftParenthesis
import com.mmust.token.NumToken
import com.mmust.token.RightParenthesis
import com.mmust.token.Token
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

class TokenizerTest {

    @Test
    @DisplayName("Arithmetic Operators Only")
    fun arithmeticOperatorsOnly() {
        val tokens = Tokenizer("3+4-5*2-9/8").parse()
        assertThat<Token>(tokens)
                .hasSize(11)
                .containsExactly(
                        NumToken("3"),
                        PLUS,
                        NumToken("4"),
                        MINUS,
                        NumToken("5"),
                        TIMES,
                        NumToken("2"),
                        MINUS,
                        NumToken("9"),
                        DIV,
                        NumToken("8")
                )
    }

    @Test
    @DisplayName("With Parenthesis")
    fun withParenthesis() {
        val tokens = Tokenizer("3*(5-1)").parse()
        assertThat<Token>(tokens)
                .hasSize(7)
                .containsExactly(
                        NumToken("3"),
                        TIMES,
                        LeftParenthesis,
                        NumToken("5"),
                        MINUS,
                        NumToken("1"),
                        RightParenthesis
                )
    }

    @Test
    @DisplayName("Without Numbers")
    fun withoutNumbers() {
        val tokens = Tokenizer("(+)").parse()

        assertThat<Token>(tokens)
                .hasSize(3)
                .containsExactly(
                        LeftParenthesis,
                        PLUS,
                        RightParenthesis
                )
    }

    @Test
    @DisplayName("Multi Digit")
    fun multiDigits() {
        val tokens = Tokenizer("125-75").parse()
        assertThat<Token>(tokens)
                .hasSize(3)
                .containsExactly(
                        NumToken("125"),
                        MINUS,
                        NumToken("75")
                )
    }

    @Test
    @DisplayName("Multi Digit Few Operators")
    fun multiDigitFewOps() {
        val tokens = Tokenizer("10*(15-3)").parse()
        assertThat<Token>(tokens)
                .hasSize(7)
                .containsExactly(
                        NumToken("10"),
                        TIMES,
                        LeftParenthesis,
                        NumToken("15"),
                        MINUS,
                        NumToken("3"),
                        RightParenthesis
                )
    }
}